hamilton/registry.py

```diff
@@ -16,6 +45,86 @@
 DATAFRAME_TYPE = "dataframe_type"
 
 
+def load_extension(plugin_module: ExtensionName):
+    """Given a module name, loads it for Hamilton to use.
+
+    :param plugin_module: the module name sans .py. e.g. pandas, polars, pyspark_pandas.
+    """
+    mod = importlib.import_module(f"hamilton.plugins.{plugin_module}_extensions")
+    # We have various plugin extensions. We default to assuming it's a dataframe extension with columns,
+    # unless it explicitly says it's not.
+    # We need to check the following if we are to enable `@extract_columns` for example.
+    extractable = getattr(mod, "COLUMN_FRIENDLY_DF_TYPE", True)
+    if extractable:
+        assert hasattr(mod, "register_types"), "Error extension missing function register_types()"
+        assert hasattr(
+            mod, f"get_column_{plugin_module}"
+        ), f"Error extension missing get_column_{plugin_module}"
+        assert hasattr(
+            mod, f"fill_with_scalar_{plugin_module}"
+        ), f"Error extension missing fill_with_scalar_{plugin_module}"
+        logger.info(f"Detected {plugin_module} and successfully loaded Hamilton extensions.")
+
+
+def initialize():
+    load_autoload_config()
+    logger.debug(f"{HAMILTON_AUTOLOAD_ENV}={os.environ.get(HAMILTON_AUTOLOAD_ENV)}")
+    for extension_name in HAMILTON_EXTENSIONS:
+        # skip modules that aren't explicitly imported by the user
+        if os.environ.get(HAMILTON_AUTOLOAD_ENV) == "0" and extension_name not in sys.modules:
```

Comment:
wont this skip importing names like polars_lazyframe ?

Reply:
actually why is the second clause needed? I thought the API is that someone has to manually register things? either via load_extension, or importing the extension which will register itself.

Reply:
Ok, maybe we have different specs. The behavior I implemented is described in the detailed in the first PR message.

### Current behavior
- When `import hamilton.function_modifiers.base`: autoload all plugins by try/except `load_extension()`
- Can't be disabled or changed

### This PR
the current behavior can be disabled to instead do the following:
- When `import hamilton.function_modifiers.base`: autoload all plugins that have been imported in `sys`. In other words,
```python
import matplotlib
from hamilton import driver
```
will import `matplotlib_extensions.py`
- after `import hamilton.function_modifiers.base`, you would need to register other extensions via `import hamilton.plugins.matplotlib_extensions` or `registry.register_extension("matplotlib")`

### The behavior you are implying
disable the current behavior can be disabled to remove *all autoloading* and only allow explicit `import hamilton.plugins.matplotlib_extensions` or `registry.register_extension("matplotlib")`

## Conclusion
I realize that the phrasing "disable autoload" might be unappropriate for this PR's behavior. Should we have 3 modes instead or would that be confusing?

1. Greedy: current behavior, default
2. Lazy: allows autoload for `import matplotlib` then `from hamilton import driver`
3. Disabled: explicit registration only

Lazy might be more valuable for development and Disabled better in prod ("disable all magic").

Side note: this is probably also why I find the CLI so slow